_posts/2019-04-24-lecture-27.md

```diff
@@ -0,0 +1,434 @@
+- Each node picks a different block, and then picks a parameter from that block to update.
+- This algorithm has a sublinear convergence.
+- Dividing the parameters into uncorrelated blocks is difficult.
+
+### Parallel Coordinate Descent with Dynamic Scheduler <d-cite key="lee2013structure"></d-cite>
+
+- STRADS (STRucture-Aware Dynamic Scheduler) is a scheduler that allows dynamic scheduling of parameter updates in coordinate descent.
+- It focuses on two things:
+  - Dependency checking: If two coordinates are correlated, do not compute them in parallel.
+  - Prioritization: Choose the coordinate to update with a probability proportional to its historic rate of change. Intuitively, a parameter that changes rapidly will probably cause a quicker decrease of the loss function.
+
+These two things together can provide an order of magnitude increase in the rate of convergence.
+
+### Advanced Optimization for Complex Constraints: Proximal Gradient methods:
+
+#### Projected Gradients:
+
+Suppose we have an optimization problem:
+
+$$
+\min_w f(w) \text{ such that } w \in C
+$$
+
+where $f$ is some smooth function and C is a convex set of points. We can define a $0-\infty$ indicator function $\iota_C$ such that:
+
+$$
+\iota_C(w) = \begin{cases}0 & w \in C\\ \infty & w \not\in C\end{cases}
+$$
+
+The optimization then reduces to:
+
+$$
+\min_w f(w) + \iota_C(w)
+$$
+
+An iterative solution can be obtained using the projected gradient descent algorithm. One iteration looks like:
+
+- $v \leftarrow w - \eta\nabla f(w)$
+- $w \leftarrow \arg\min_z \frac{1}{2\eta}\lVert v - z\rVert_2^2 + \iota_C(z) = \arg\min_{z \in C}\frac{1}{2\eta}\lVert v - z\rVert_2^2$
+
+Intuitively, it is equivalent to taking an unconstrained step in the direction of the gradient, and then projecting the new position to the closest point in the feasible set.
+
+#### Proximal Gradients:
+
+Proximal gradients are a generalization of projected gradients, in which the indicator function $\iota_C$ is replaced by an arbitrary function $g(z)$, which may model constraints, regularizations etc. Intuitively, now, not only can we restrict $w$ to a feasible region, but can also say which regions are better than the others. Now one iteration of the algorithm looks like:
+
+- $v \leftarrow w - \eta\nabla f(w)$
+- $w \leftarrow \arg\min_z \frac{1}{2\eta}\lVert v - z\rVert_2^2 + g(z)$
+
+The minimization problem in the second step has been well-studied, and closed form operators have been derived for the expression $\arg\min_z \frac{1}{2\eta}\lVert v - z\rVert_2^2 + g(v)$ for different $g$. We denote:
+
+$$
+P^\eta_g(v) = \arg\min_z \frac{1}{2\eta}\lVert v - z\rVert_2^2 + g(z)
+$$
+
+Also, for certain conditions derived by <d-cite key="yu2013decomposing"></d-cite>, $P^\eta_{g_1+g_2}(v) = p^\eta_{g_1}(p^\eta_{g_2}(v))$, allowing us to model combinations of constraints/regularizations. 
+
+#### Accelerated Proximal Gradients:
```

Comment:
Please format this section in paragraph form so that it reads more like a blog post, rather than bullet points from lecture slides. Here's an example that I liked: https://sailinglab.github.io/pgm-spring-2019/notes/lecture-13/